src/common/base/src/range_read.rs

```diff
@@ -80,26 +88,162 @@ impl<R: RangeReader + Send + Unpin> RangeReader for &mut R {
     }
 }
 
-/// `RangeReaderAdapter` bridges `RangeReader` and `AsyncRead + AsyncSeek`.
-pub struct RangeReaderAdapter<R>(pub R);
+/// `AsyncReadAdapter` adapts a `RangeReader` to an `AsyncRead`.
+#[pin_project]
+pub struct AsyncReadAdapter<R> {
+    /// The inner `RangeReader`.
+    /// Use `Mutex` to get rid of the borrow checker issue.
+    inner: Arc<Mutex<R>>,
+
+    /// The current position from the view of the reader.
+    position: u64,
+
+    /// The buffer for the read bytes.
+    buffer: Vec<u8>,
+
+    /// The length of the content.
+    content_length: u64,
+
+    /// The future for reading the next bytes.
+    #[pin]
+    read_fut: Option<Pin<Box<dyn Future<Output = io::Result<Bytes>> + Send>>>,
+}
+
+impl<R: RangeReader + 'static> AsyncReadAdapter<R> {
+    pub async fn new(inner: R) -> io::Result<Self> {
+        let mut inner = inner;
+        let metadata = inner.metadata().await?;
+        Ok(AsyncReadAdapter {
+            inner: Arc::new(Mutex::new(inner)),
+            position: 0,
+            buffer: Vec::new(),
+            content_length: metadata.content_length,
+            read_fut: None,
+        })
+    }
+}
+
+impl<R: RangeReader + 'static> AsyncRead for AsyncReadAdapter<R> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let mut this = self.as_mut().project();
+
+        if *this.position >= *this.content_length {
+            return Poll::Ready(Ok(0));
+        }
+
+        if !this.buffer.is_empty() {
+            let to_read = this.buffer.len().min(buf.len());
+            buf[..to_read].copy_from_slice(&this.buffer[..to_read]);
+            this.buffer.drain(..to_read);
+            *this.position += to_read as u64;
+            return Poll::Ready(Ok(to_read));
+        }
+
+        if this.read_fut.is_none() {
+            let range = *this.position..*this.content_length;
+            let inner = this.inner.clone();
+            let fut = async move {
+                let mut inner = inner.lock().await;
+                inner.read(range).await
+            };
```

Comment:
If the content_length is too large, reading the whole range may cause oom. Maybe we should limit the max size to read.

Reply:
added